### DISCUSSION TITLE
Relation / Many To Many question

### DISCUSSION THREAD

**Discussion by elbart** [SeaQL/sea-orm#278] [Q&A] [answered]
Hi all,

I am having a quite hard time to actually get a many to many relation selected and at least based on my limited experience the query you provide as an example in your docs also looks wrongish. I'd be happy if you have any guidance for me.

My project does not have Cakes and Fillings, but Recipes and Ingredients. The Join Table in between is called recipe_ingredient. Apart from this - same same. But let's get back to your example in: https://www.sea-ql.org/SeaORM/docs/advanced-query/advanced-relations#custom-joins

Here you explain how this works and I tried this for my entities too and the query is generated in the same way as yours. However the generated join part of the query seems super *wrong* (at least for postgres - postgres complains about `cake` to be used twice):

```bash
"SELECT `cake`.`id`, `cake`.`name`, FROM `cake`",
"INNER JOIN `cake` ON `cake_filling`.`cake_id` = `cake`.`id`",
"INNER JOIN `filling` ON `cake_filling`.`filling_id` = `filling`.`id`",
```

My assumption of a valid query would be (the second join is joining `cake` again, not `cake_filling`):

```bash
"SELECT `cake`.`id`, `cake`.`name`, FROM `cake`",
"INNER JOIN `cake_filling` ON `cake_filling`.`cake_id` = `cake`.`id`",
"INNER JOIN `filling` ON `cake_filling`.`filling_id` = `filling`.`id`",
```

a) Am I wrong? 
b) Is that simply fixable and if yes, how?

Thank you for your support!

**Comment by billy1624** [comment, marked as answer]
Hi @elbart, thanks for pointing that out! The [example](https://www.sea-ql.org/SeaORM/docs/advanced-query/advanced-relations#custom-joins) is horribly wrong... to fix it uses `join_rev` instead.

```rust
assert_eq!(
    cake::Entity::find()
        .column_as(filling::Column::Id.count(), "count")
        .join_rev(
            // construct `RelationDef` on the fly
            JoinType::InnerJoin,
            cake_filling::Entity::belongs_to(cake::Entity)
                .from(cake_filling::Column::CakeId)
                .to(cake::Column::Id)
                .into()
        )
        // reuse a `Relation` from existing Entity
        .join(JoinType::InnerJoin, cake_filling::Relation::Filling.def())
        .group_by(cake::Column::Id)
        .having(filling::Column::Id.count().equals(Expr::value(2)))
        .build(DbBackend::MySql)
        .to_string(),
    [
        "SELECT `cake`.`id`, `cake`.`name`, COUNT(`filling`.`id`) AS `count` FROM `cake`",
        "INNER JOIN `cake_filling` ON `cake_filling`.`cake_id` = `cake`.`id`",
        "INNER JOIN `filling` ON `cake_filling`.`filling_id` = `filling`.`id`",
        "GROUP BY `cake`.`id`",
        "HAVING COUNT(`filling`.`id`) = 2",
    ]
    .join(" ")
);
```

**Comment by elbart** [comment]
Thank you very much! That worked indeed!

**Comment by tyt2y3** [comment]
Thank you! @billy1624
